README.md
# gotorsocks

**gotorsocks** - Easy to use Golang's **torify**. Provides CLI access to **Tor** proxying inside of **Go apps**. So, can parse web-sites on the 
**.onion domens**.

It is GitHub **fork** of [this BitBucket's repo](https://bitbucket.org/kallevedin/torsocks).

Old import path "code.google.com/p/go.net/proxy" (same as "golang.org/x/net/proxy") used in original version, is deadly broken, so original 
package uninstallable. In current version import path corrected, and some detail of code is changed.

[![Go Report Card](https://goreportcard.com/badge/github.com/hIMEI29A/gotorsocks)](https://goreportcard.com/report/github.com/hIMEI29A/gotorsocks) [![GoDoc](https://godoc.org/github.com/hIMEI29A/gotorsocks?status.svg)](http://godoc.org/github.com/hIMEI29A/gotorsocks)



## Install

Project uses [Glide](https://glide.sh) for dependencies manage. 
So, **install Glide first**:

```sh
curl https://glide.sh/get | sh
```
**Clone** repo:

```sh
git clone https://github.com/hIMEI29A/gotorsocks.git

cd gotorsocks
```

**Update** deps and install:

```sh
glide update

glide install
```
It will install dependencies ("golang.org/x/net/proxy" only) to `vendor/` folder of the repo.

**Run**

```sh
make
```

It will **run tests** and if it passed, install **gotorsocks** in to your $GOPATH

Also you can simply

```sh
go get github.com/hIMEI29A/gotorsocks
```

Check the [release page](https://github.com/hIMEI29A/gotorsocks/releases)!

## Example

```go
package main

import (
    "bufio"
    "fmt"

    "github.com/hIMEI29A/gotorsocks"
)

func main() {
    address := "facebookcorewwwi.onion:80"
    tor, err := gotorsocks.NewTorGate()

    if err != nil {
        panic(err)
    }

    connect, err := tor.DialTor(address)

    if err != nil {
        panic(err)
    }

    fmt.Fprintf(connect, "GET / HTTP/1.0\r\n\r\n")
    status, err := bufio.NewReader(connect).ReadString('\n')
    fmt.Println(status)
}
```


CHANGELOG.md
# Changelog - gotorsocks

### 0.3.2

- Makefile's error fixed
- Main file gotorsocks.go changed

### 0.2.1

__Changes__

- pre-releasese commit; previous version was unmaintained, untagged and unreleased
- initial commit

__Contributors__

- hIMEI

Released by hIMEI, Sat 16 Dec 2017 -
[see the diff](https://github.com/<no value>/torsocks/compare/13e646ac6275441bc2ff41f0de64b44653cb8384...0.2.1#diff)
______________




TorSOCKS_test.go
package torsocks

import (
	"testing"
	"fmt"
	)



func TestBadProxy(t *testing.T) {
	fmt.Println("Testing bad Tor Proxy")
	_, err := NewTorProxy("example.com:22")
	if err == nil {
		t.Fail()
	}
}


func TestBadConnect(t *testing.T) {
	fmt.Println("Testing to connect through 127.0.0.1:9050, to herp.derp:80.")
	tp, err := NewTorProxy("127.0.0.1:9050")
	if err != nil {
		fmt.Println("Maybe you are not running Tor?")
		fmt.Println(err.Error())
		t.Fail()
	}
	_, err = tp.DialTor("herp.derp:80")
	if err == nil {
		fmt.Println("We actually managed to connect to herp.derp. This is odd, and probably WRONG.\n")
		fmt.Println(err.Error())
		t.Fail()
	}
}


func TestConnect(t *testing.T) {
	onion := "227vftpsbp62v7bd.onion:6667"
	fmt.Println("Testing to connect through 127.0.0.1:9050, to " + onion)

	tp, err := NewTorProxy("127.0.0.1:9050")
	if err != nil {
		fmt.Println("Maybe you are not running Tor?")
		fmt.Println(err.Error())
		t.Fail()
		return
	}
	fmt.Printf("Dialing to " + onion + "\n")
	_, err = tp.DialTor(onion)
	if err != nil {
		fmt.Println("Maybe spore.mcx (" + onion + ") is not listening. Error was: " + err.Error())
		fmt.Println(err.Error())
		t.Fail()
		return
	}
	fmt.Println("Connected to .onion successfully!")
}



torsocks_test.go
package torsocks

import (
	"fmt"
	"testing"
)

func TestBadConnect(t *testing.T) {
	fmt.Println("Testing to connect through 127.0.0.1:9050, to herp.derp:80.")
	tp, err := NewTorGate()
	if err != nil {
		fmt.Println("Maybe you are not running Tor?")
		fmt.Println(err.Error())
		t.Fail()
	}
	_, err = tp.DialTor("herp.derp:80")
	if err == nil {
		fmt.Println("We actually managed to connect to herp.derp. This is odd, and probably WRONG.\n")
		fmt.Println(err.Error())
		t.Fail()
	}
}

func TestConnect(t *testing.T) {
	onion := "facebookcorewwwi.onion:80"
	fmt.Println("Testing to connect through 127.0.0.1:9050, to " + onion)

	tp, err := NewTorGate()
	if err != nil {
		fmt.Println("Maybe you are not running Tor?")
		fmt.Println(err.Error())
		t.Fail()
		return
	}
	fmt.Printf("Dialing to " + onion + "\n")
	_, err = tp.DialTor(onion)
	if err != nil {
		fmt.Println("Maybe spore.mcx (" + onion + ") is not listening. Error was: " + err.Error())
		fmt.Println(err.Error())
		t.Fail()
		return
	}
	fmt.Println("Connected to .onion successfully!")
}


gotorsocks_test.go
package gotorsocks

import (
	"fmt"
	"testing"
)

func TestBadConnect(t *testing.T) {
	fmt.Println("Testing to connect through 127.0.0.1:9050, to herp.derp:80.")
	tp, err := NewTorGate()
	if err != nil {
		fmt.Println("Maybe you are not running Tor?")
		fmt.Println(err.Error())
		t.Fail()
	}
	_, err = tp.DialTor("herp.derp:80")
	if err == nil {
		fmt.Println("We actually managed to connect to herp.derp. This is odd, and probably WRONG.")
		fmt.Println(err.Error())
		t.Fail()
	}
}

func TestConnect(t *testing.T) {
	onion := "facebookcorewwwi.onion:80"
	fmt.Println("Testing to connect through 127.0.0.1:9050, to " + onion)

	tp, err := NewTorGate()
	if err != nil {
		fmt.Println("Maybe you are not running Tor?")
		fmt.Println(err.Error())
		t.Fail()
		return
	}
	fmt.Printf("Dialing to " + onion + "\n")
	_, err = tp.DialTor(onion)
	if err != nil {
		fmt.Println("Maybe spore.mcx (" + onion + ") is not listening. Error was: " + err.Error())
		fmt.Println(err.Error())
		t.Fail()
		return
	}
	fmt.Println("Connected to .onion successfully!")
}


torsocks.go
// @Package torsocks
// @Author: Kalle Vedin <kalle.vedin@fripost.org>
// @Author: hIMEI <himei@tuta.io>
// @Date:   2017-12-16 22:02:59
// @Copyright © 2017 hIMEI <himei@tuta.io>
// @license MIT

// Here is is GitHub fork of https://bitbucket.org/kallevedin/torsocks.
// Import path "code.google.com/p/go.net/proxy" (same as "golang.org/x/net/proxy")
// used in original version, is fatal broken, so packege uninstallabnle.
// In current version import path corrected, and some detail of code is changed
//
// https://bitbucket.org/kallevedin/torsocks is relised on Public Domain.

package torsocks

import (
	"bytes"
	"errors"
	"net"
	"time"

	"golang.org/x/net/proxy"
)

// A Tor proxy. Is actually just a string with the address of the Tor Proxy. (Needs to be an IPv4
// address or a domain name that can be translated to an IPv4 address, with a port.)
// Examples: "127.0.0.1:9050", "10.0.30.11:9150".
type TorGate string

// String constant with localhost's Tor port
const TOR_GATE_ string = "127.0.0.1:9050"

// Creates new TorGate
func NewTorGate() (*TorGate, error) {
	duration, _ := time.ParseDuration("10s")
	connect, err := net.DialTimeout("tcp4", TOR_GATE_, duration)

	if err != nil {
		return nil, errors.New("Could not test TOR_GATE_: " + err.Error())
	}

	// Tor proxies reply to anything that looks like
	// HTTP GET or POST with known error message.
	connect.Write([]byte("GET /\n"))
	connect.SetReadDeadline(time.Now().Add(10 * time.Second))
	buf := make([]byte, 4096)

	for {
		n, err := connect.Read(buf)

		if err != nil {
			return nil, errors.New("It is not TOR_GATE_")
		}

		if bytes.Contains(buf[:n], []byte("Tor is not an HTTP Proxy")) {
			connect.Close()
			gate := TorGate(TOR_GATE_)

			return &gate, nil
		}
	}
}

// Dials to the .onion address
func (gate *TorGate) DialTor(address string) (net.Conn, error) {
	dialer, err := proxy.SOCKS5("tcp4", string(*gate), nil, proxy.Direct)

	if err != nil {
		return nil, errors.New("Could not connect to TOR_GATE_: " + err.Error())
	}

	connect, err := dialer.Dial("tcp4", address)

	if err != nil {
		return nil, errors.New("Failed to connect: " + err.Error())
	}

	return connect, nil
}


gotorsocks.go
// @Package gotorsocks
// @Author: Kalle Vedin <kalle.vedin@fripost.org>
// @Author: hIMEI <himei@tuta.io>
// @Date:   2017-12-16 22:02:59
// @Copyright © 2017 hIMEI <himei@tuta.io>
// @license MIT

// gotorsocks Here is GitHub fork of https://bitbucket.org/kallevedin/torsocks.
// Import path "code.google.com/p/go.net/proxy" (same as "golang.org/x/net/proxy")
// used in original version, is fatal broken, so original package uninstallable.
// In current version import path corrected, and some detail of code is changed
//
// https://bitbucket.org/kallevedin/torsocks is relised on Public Domain.

package gotorsocks

import (
	"bytes"
	"errors"
	"net"
	"time"

	"golang.org/x/net/proxy"
)

// TorGate is  a Tor proxy. Is actually just a string with the address of the Tor Proxy.
// (Needs to be an IPv4 address or a domain name that can be translated to an IPv4
// address, with a port.)
// Examples: "127.0.0.1:9050", "10.0.30.11:9150".
type TorGate string

// TOR_GATE string constant with localhost's Tor port
const TOR_GATE_ string = "127.0.0.1:9050"

// NewTorGate creates new TorGate
func NewTorGate() (*TorGate, error) {
	duration, _ := time.ParseDuration("10s")
	connect, err := net.DialTimeout("tcp4", TOR_GATE_, duration)

	if err != nil {
		return nil, errors.New("Could not test TOR_GATE_: " + err.Error())
	}

	// Tor proxies reply to anything that looks like
	// HTTP GET or POST with known error message.
	connect.Write([]byte("GET /\n"))
	connect.SetReadDeadline(time.Now().Add(10 * time.Second))
	buf := make([]byte, 4096)

	for {
		n, err := connect.Read(buf)

		if err != nil {
			return nil, errors.New("It is not TOR_GATE_")
		}

		if bytes.Contains(buf[:n], []byte("Tor is not an HTTP Proxy")) {
			connect.Close()
			gate := TorGate(TOR_GATE_)

			return &gate, nil
		}
	}
}

// DialTor dials to the .onion address
func (gate *TorGate) DialTor(address string) (net.Conn, error) {
	dialer, err := proxy.SOCKS5("tcp4", string(*gate), nil, proxy.Direct)

	if err != nil {
		return nil, errors.New("Could not connect to TOR_GATE_: " + err.Error())
	}

	connect, err := dialer.Dial("tcp4", address)

	if err != nil {
		return nil, errors.New("Failed to connect: " + err.Error())
	}

	return connect, nil
}


TorSOCKS.go
// Here is is GitHub fork of https://bitbucket.org/kallevedin/torsocks.
// Import path "code.google.com/p/go.net/proxy" (same as "golang.org/x/net/proxy")
// used in original version, is fatal broken, so packege uninstallabnle.
// In current version import path corrected, and some detail of code is changed
//
// https://bitbucket.org/kallevedin/torsocks is relised on Public Domain.

package torsocks

import (
	"bytes"
	"errors"
	"net"
	"time"

	"golang.org/x/net/proxy"
)

// A Tor proxy. Is actually just a string with the address of the Tor Proxy. (Needs to be an IPv4
// address or a domain name that can be translated to an IPv4 address, with a port.)
// Examples: "127.0.0.1:9050", "10.0.30.11:9150".
type TorProxy string

// Defines a new Tor SOCKS5 proxy. Also performs very basic heuristic check to see if it actually
// is a Tor proxy we are connecting to, and not a normal SOCKS proxy, or something else.
func NewTorProxy(address string) (*TorProxy, error) {
	duration, _ := time.ParseDuration("10s")
	conn, err := net.DialTimeout("tcp4", address, duration)
	if err != nil {
		return nil, errors.New("Could not test Tor Proxy: " + err.Error())
	}
	// Tor proxies reply to anything that looks like HTTP GET or POST with known error message.
	conn.Write([]byte("GET /\n"))
	conn.SetReadDeadline(time.Now().Add(10 * time.Second))
	buf := make([]byte, 4096)
	for {
		n, err := conn.Read(buf)
		if err != nil {
			return nil, errors.New("Failed heuristics. " + address + " is maybe not a Tor Proxy.")
		}
		if bytes.Contains(buf[:n], []byte("Tor is not an HTTP Proxy")) {
			conn.Close()
			tp := TorProxy(address)
			return &tp, nil
		}
	}
}

// Defines a new Tor SOCKS5 proxy. Does not perform any heuristic checks, and always succeeds.
func NewTorProxyNoCheck(address string) *TorProxy {
	t := TorProxy(address)
	return &t
}

// Dials to the address, through the Tor Proxy. address needs to be an IPv4 address or a domain name
// that can be translated to an IPv4 address (or a .onion address!) followed by the port number you
// want to connect to. Examples: "1.2.3.4:22", "duckduckgo.com:80", "227vftpsbp62v7bd.onion:6667".
func (tp *TorProxy) DialTor(address string) (net.Conn, error) {
	dialer, err := proxy.SOCKS5("tcp4", string(*tp), nil, proxy.Direct)
	if err != nil {
		return nil, errors.New("Could not connect to Tor proxy SOCKS: " + err.Error())
	}
	conn, err := dialer.Dial("tcp4", address)
	if err != nil {
		return nil, errors.New("Failed to connect: " + err.Error())
	}
	return conn, nil
}
